### PR TITLE
Deleted entity access

### PR DESCRIPTION
**Pull request by swilly22**
Allow read-only access to deleted entities

### CODE REVIEW

File: src/graph/entities/attribute_set.c [hunk 1]
```diff
@@ -9,12 +9,15 @@
...
 // compute size of attribute set in bytes
-#define ATTRIBUTESET_BYTE_SIZE(set) ((set) == NULL ? sizeof(_AttributeSet) : sizeof(_AttributeSet) + sizeof(Attribute) * (set)->attr_count)
+#define ATTRIBUTESET_BYTE_SIZE(set) ((set) == NULL ? \
+		sizeof(_AttributeSet) :                      \
+		sizeof(_AttributeSet) + sizeof(Attribute) * (set)->attr_count)
 
-// determine if set is empty
-#define ATTRIBUTESET_EMPTY(set) (set) == NULL
+// mark attribute-set as mutable
+#define ATTRIBUTE_SET_CLEAR_MSB(set) ((intptr_t)(set) & MSB_MASK_CMP)
```
Comment (AviAvni) [inline src/graph/entities/attribute_set.c:20]: use CLEAR_MSB from RG.h

File: src/graph/entities/attribute_set.c [hunk 2]
```diff
@@ -284,12 +340,21 @@ bool AttributeSet_Update
...
 	ASSERT(attr_id != ATTRIBUTE_ID_NONE);
 
+	AttributeSet _set = *set;
+
+	// return if set is read-only
+	if(unlikely(ATTRIBUTE_SET_IS_READONLY(_set))) {
+		return false;
+	}
+
+	ASSERT(_set != NULL);
```
Comment (AviAvni) [inline src/graph/entities/attribute_set.c:350]: remove assert

File: src/graph/entities/attribute_set.c [hunk 3]
```diff
@@ -353,6 +399,11 @@ void AttributeSet_PersistValues
 (
 	const AttributeSet set  // set to persist
 ) {
+	// return if set is read-only
+	if(unlikely(ATTRIBUTE_SET_IS_READONLY(set))) {
```
Comment (AviAvni) [inline src/graph/entities/attribute_set.c:403]: can we make this assert?

File: src/graph/graph.c [hunk 4]
```diff
@@ -1159,6 +1159,11 @@ inline bool Graph_EntityIsDeleted
 (
 	const GraphEntity *e
 ) {
+	if(e->attributes == NULL) {
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/graph/graph.c:1162]: I can't remove this,
Consider `CREATE (a)-[:R]->(b)`
Where both `a` and `b` are reserved nodes, which means their internal attributes pointer is set to `NULL`.

As part of the **NEW** edge creation logic we want to make sure we're not creating an edge where one of its ends (src / dest) isn't a deleted node. to support this check for all "types" of nodes:
1. binded
2. reserved
3. deleted

requires the addition of this `if` condition.

File: src/graph/entities/attribute_set.c (hunk 2 above)
Comment (swilly22) [reply src/graph/entities/attribute_set.c:350]: going to keep it, it seems like the only place which calls this function make sure the updated attribute exists, also I think update should acts on an existing attribute, we can't update a none existing attribute

File: src/graph/entities/attribute_set.c (hunk 3 above)
Comment (swilly22) [reply src/graph/entities/attribute_set.c:403]: yes we can!

**Pull request merged**